topics/functions-intro.md

```diff
@@ -0,0 +1,353 @@
+On the other hand, we do want to provide the means to obtain the modification timestamp for a given Hash key.
+
+We'll add two new functions to our library to accomplish these objectives:
+
+1. The `my_hgetall` Redis Function will return all fields and their respective values from a given Hash key name, excluding the metadata (i.e., the _\_last_updated\__ field).
+1. The `my_hlastmodified` Redis Function will return the modification timestamp for a given Hash key name.
+
+The library's source code could look something like the following:
+
+```lua
+local function my_hset(keys, args)
+  local hash = keys[1]
+  local time = redis.call('TIME')[1]
+  return redis.call('HSET', hash, '_last_modified_', time, unpack(args))
+end
+
+local function my_hgetall(keys, args)
+  redis.setresp(3)
+  local hash = keys[1]
+  local res = redis.call('HGETALL', hash)
+  res['map']['_last_modified_'] = nil
+  return res
+end
+
+local function my_hlastmodified(keys, args)
+  local hash = keys[1]
+  return redis.call('HGET', keys[1], '_last_modified_')
+end
+
+redis.register_function('my_hset', my_hset)
+redis.register_function('my_hgetall', my_hgetall)
+redis.register_function('my_hlastmodified', my_hlastmodified)
+```
+
+While all of the above should straightforward, note that the `my_hgetall` also calls [`redis.setresp(3)`](/topics/lua-api#redis.setresp).
+That means that the function expects [RESP3](https://github.com/redis/redis-specifications/blob/master/protocol/RESP3) replies after calling `redis.call()`, which, unlike the default RESP2 protocol, provides dictionary (associative arrays) replies.
+Doing so allows the function to delete (or set to `nil` as is the case with Lua tables) specific fields from the reply, and in our case, the _\_last_modified\__ field.
+
+Assuming you've saved the library's implementation in the _mylib.lua_ file, you can replace it with its (optional) description with:
+
+```bash
+$ cat mylib.lua | redis-cli -x FUNCTION LOAD Lua mylib REPLACE DESCRIPTION "My application's Hash data type enhancements"
+```
+
+Once loaded, you can call the library's functions with `FCALL`:
+
+```
+redis> FCALL my_hgetall 1 myhash
+1) "myfield"
+2) "some value"
+3) "another_field"
+4) "another value"
+redis> FCALL my_hlastmodified 1 myhash
+"1640772721"
+```
+
+You can also get the library's details with the `FUNCTION LIST` command:
+
+```
+redis> FUNCTION LIST
+1) 1) "library_name"
+   2) "mylib"
+   3) "engine"
+   4) "LUA"
+   5) "description"
+   6) "My application's Hash data type enhancements"
+   7) "functions"
+   8) 1) 1) "name"
+         2) "my_hset"
+         3) "description"
+         4) (nil)
+      2) 1) "name"
+         2) "my_hgetall"
+         3) "description"
+         4) (nil)
+      3) 1) "name"
+         2) "my_hlastmodified"
+         3) "description"
+         4) (nil)
+```
+
+You can see that it is easy to update our library with new capabilities.
+
+## Reusing code in the library
```

Comment:
we're missing a section about flags.
maybe add it in a follow up PR (so add a todo comment at the top or something).
when we do add it, keep in mind that the're soon gonna serve EVAL too.